### PR TITLE
Adds additing string ordering in the query tool

### PR DESCRIPTION
**Pull request by SaptakS**
Before submitting pull request, please ensure that:

- [x] Every commit has a message which describes it
- [x] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them

Adds the following sortings:
- Labels
- Age of string
- Number of failing checks
- Number of comments

Refs #3710

### CODE REVIEW

File: weblate/trans/models/unit.py [hunk 1]
```diff
@@ -118,16 +118,42 @@ def same(self, unit, exclude=True):
...
-        sort_list = [
-            choice
-            for choice in sort_list_request
-            if choice.replace("-", "") in available_sort_choices
+        available_sort_choices = [
+            "priority",
+            "position",
+            "context",
+            "num_words",
+            "labels",
```
Comment (nijel) [inline weblate/trans/models/unit.py:126]: This should probably translate to something like `labels__name`, but it needs to be tested.

Comment (SaptakS) [reply weblate/trans/models/unit.py:126]: Correct. I will test this out.

File: weblate/trans/tests/test_models.py [hunk 2]
```diff
@@ -334,6 +335,37 @@ def test_flags(self):
...
+
+        # test sorting by count
+        unit4 = Unit.objects.filter(translation__language_code="cs")[2]
+        Comment.objects.create(unit=unit4, comment="Foo")
+        unit5 = Unit.objects.filter(translation__language_code="cs")
+        unit5 = unit5.order_by_request({"sort_by": "-num_comments"})
+        self.assertEqual(unit5[0].comment_set.count(), 1)
+        unit5 = Unit.objects.filter(translation__language_code="cs")
+        unit5 = unit5.order_by_request({"sort_by": "num_comments"})
+        self.assertEqual(unit5[0].comment_set.count(), 0)
```
Comment (nijel) [inline weblate/trans/tests/test_models.py:367]: Can you please test all ordering options? At least to the level that the generated queries can be executed, it doesn't have to cover validation of the ordering being correct.

File: weblate/trans/models/unit.py [hunk 3]
```diff
@@ -118,16 +118,45 @@ def same(self, unit, exclude=True):
...
+            unsigned_choice = choice.replace("-", "")
+            if unsigned_choice in countable_sort_choices:
+                return self.order_by_count(
+                    choice.replace(
+                        unsigned_choice,
+                        countable_sort_choices[unsigned_choice]["order_by"],
+                    ),
+                    countable_sort_choices[unsigned_choice]["filter"],
+                )
+            elif unsigned_choice in available_sort_choices:
```
Comment (nijel) [inline weblate/trans/models/unit.py:147]:
```suggestion
            if unsigned_choice in available_sort_choices:
```
No need for else after return.

File: weblate/trans/models/unit.py [hunk 4]
```diff
@@ -118,16 +118,45 @@ def same(self, unit, exclude=True):
...
+                if unsigned_choice == "labels":
+                    self = self.annotate(max_labels_name=Max("labels__name"))
```
Comment (nijel) [inline weblate/trans/models/unit.py:149]: Please do not assign to self.

File: weblate/trans/tests/test_models.py (hunk 2 above)
Comment (SaptakS) [reply weblate/trans/tests/test_models.py:367]: Have added a simple test to loop over all the ordering options and check if the queryset count still works. I hope that kind of ensures what you were asking. Also, I have added a similar test to check that multiple ordering options also doesn't cause any error. Let me know if this seems good.

File: weblate/trans/tests/test_models.py [hunk 5]
```diff
@@ -366,6 +366,28 @@ def test_order_by_request(self):
...
+            "num_words",
+            "labels",
+            "timestamp",
+            "num_failing_checks",
+        ]
+        for order_option in order_options:
+            ordered_unit = Unit.objects.filter(
+                translation__language_code="cs"
+            ).order_by_request({"sort_by": order_option})
+            self.assertEqual(ordered_unit.count(), 4)
```
Comment (nijel) [inline weblate/trans/tests/test_models.py:383]:
```suggestion
            self.assertEqual(len(ordered_unit), 4)
```

I'm not sure if `count` doesn't discard ordering (it has no use for it), so better let the query evaluate fully.

File: weblate/trans/tests/test_models.py [hunk 6]
```diff
@@ -380,7 +380,7 @@ def test_order_by_request(self):
             ordered_unit = Unit.objects.filter(
                 translation__language_code="cs"
             ).order_by_request({"sort_by": order_option})
-            self.assertEqual(ordered_unit.count(), 4)
+            self.assertEqual(len(ordered_unit), 4)
```
Comment (SaptakS) [inline weblate/trans/tests/test_models.py:383]: Aah. I see what you mean. Makes sense.

**Comment by nijel** [comment]
Merged, thanks for your contribution!

**Pull request merged**